### ISSUE TITLE
watch for changes to configmap/secret and trigger App reconcile

### ISSUE THREAD

**Issue by cppforlife** [carvel-dev/kapp-controller#26]
this would make syncPeriod configuration more useful (e.g. it could be set higher, and App can try to react to "input" changes proactively). i'd like to have this mechanims pave the way for "watching" of other resources used by App.

**Comment by cppforlife** [comment]
current thinking:

```yaml
apiVersion: kappctrl.k14s.io/v1alpha1
kind: AppReconcileRequest
metadata:
  generateName: app-name-
spec:
  appRef:
    name: app-name
  reason: "ConfigMap x contents changed."
status:
  conditions:
  - type: Enqueued
```

controller would delete AppReconcileRequest after X minutes after it was enqueued.

alternatively, we could add some field like `spec.reconcileMarker` on `App` resource to force metadata.generation to be bumped, hence would reconcile. downside to that is it makes App's spec possibly conflict with what user is entering (if we do not add kapp rebase rules for that) and less declarative.

**Comment by loewenstein** [comment]
Who would _create_ those `AppReconcileRequest` CRs?

**Comment by cppforlife** [comment]
kapp-controller itself (when it's watching for configmap/secrets changes) or other controllers when they for some reason want to force app reconciliation.

**Comment by loewenstein** [comment]
@cppforlife mmh, what about the kapp-controller just reconciling (instead of creating a CR, observe the CR and then reconciling)?
Forced reconciliation, by human operator or controller, could be requested nicely via annotation.
WDYT?

**Comment by cppforlife** [comment]
> Forced reconciliation, by human operator or controller, could be requested nicely via annotation.

annotations do not force metadata.generation which makes it weird for reconciliation. it also feels a bit weird when multiple entities need to trigger change.

> what about the kapp-controller just reconciling (instead of creating a CR, observe the CR and then reconciling)?

id like to consider adding additional controllers at some point that might be more proactive around external changes (e.g. controller that watches git repo / accepts webhooks etc.). very fuzzy on details.

**Comment by danielhelfand** [comment]
This has been added in #127. Just waiting on release to take place for `v0.17.0`.

**Comment by jessehu** [comment]
Hi @cppforlife @danielhelfand is it possible to watch for delete events of some resources created by App CR? The use case is users manually delete StorageClass CR, but kapp-controller will wait for syncPeriod to create it again.

**Comment by 100mik** [comment]
@jessehu would you mind creating another issue?

This one does mention introduction of mechanisms for watching resources, however it is more concerned with watching for changes in "inputs" to the App.
Looks like your case is more focussed on responding faster to events on certain critical resources created by kapp-controller, correct?

A separate issue will help others with similar use cases chime in.

**Comment by jessehu** [comment]
Thanks @100mik. Sorry for the late response. I submitted a feature request on https://github.com/carvel-dev/kapp-controller/issues/1279
